### DISCUSSION TITLE
Optimal Label Encoding for YOLOv8 Segmentation from Raw Contours and Analyzing Confusion Matrix True Negative Discrepancies

### DISCUSSION THREAD

**Discussion by MarioPasc** [ultralytics/ultralytics#9202] [Q&A]
Greetings,

I am currently engaged in a project focused on medical image segmentation for my BSc thesis. In the last three weeks, I have been exploring the application of YOLOv8 for this purpose. I have two primary inquiries:

1. Regarding the conversion of raw Region of Interest (ROI) images (.nii format) into labels suitable for YOLOv8 segmentation tasks, would the following approach be considered appropriate? (I only have one class: tumor)

```python
def contours_YOLO_format(contours, height, width, output_path):
    if len(contours) > 0:  # There are contours in the ROI
        if math.ceil(len(contours[0]) / 2) >= 3: # There are at least 3 tuples <x, y> in the ROI: 6 points minimum
            with open(output_path, 'w') as f:
                for contour in contours:
                    normalized = contour.squeeze().astype('float') / np.array([width, height])
                    str_contour = ' '.join([f"{coord:.6f}" for coord in normalized.flatten()])
                    f.write(f"0 {str_contour}\n")
```

Contours have been computed using the `openCV` function: 
```python
contours, _ = cv.findContours(data.astype(np.uint8), cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
```
Where height and width are:
```python
img = nib.load(os.path.join(nii_folder, nii_file))
data = img.get_fdata()

width = data.shape[1]
height = data.shape[0]
```

2. I am encountering a challenge similar to the one discussed in [this issue](https://github.com/ultralytics/ultralytics/issues/4960#issuecomment-1957959843) on your GitHub page. There is a discrepancy between the F1-confidence plots and the confusion matrix, indicating that the inclusion of empty images (no ROI) might be affecting certain aspects of the validation process. Specifically, the True Negative value (True: Background, Predicted: Background) consistently appears as blank. ¿Is there any solution to this?. 

```python
model = YOLO(model=weight_file, task='segment')
results = model.val(data=self.yaml_file, imgsz=640, conf=0.01, plots=True)

conf_mat = results.confusion_matrix.matrix
TP = conf_mat[0][0]
TN = conf_mat[1][1]
FP = conf_mat[0][1]
FN = conf_mat[1][0]
```
The TN value returned by `ultralytics.utils.metrics.ConfusionMatrix` is also 0. 

I appreciate your time and look forward to any guidance you can provide.

**Comment by pderrenger** [comment]
@MarioPasc hello,

Your inquiries are spot on, and I'm delighted to assist with them. 🚀

1. **Optimal Label Encoding for Segmentation**:
Your approach to converting ROI contours to the YOLOv8 segmentation label format seems reasonable given you're working with tumor segmentation in medical images. Remember, YOLOv8 excels with normalized polygon points for segmenting objects rather than traditional bounding boxes, especially in complex shapes such as tumors. Your method of normalizing contour points and writing them to a file perfectly aligns with what is needed for YOLOv8 segmentation tasks.

2. **Discrepancies in Confusion Matrix**: 
For segmentation tasks with YOLOv8, images with no Region of Interest (ROI) can indeed introduce challenges in evaluating model performance, particularly affecting True Negative (TN) counts. Unlike object detection, segmentation models like YOLOv8 often don't explicitly predict "background" as a class; instead, they focus on the foreground or areas of interest (e.g., tumors). This can lead to confusion about what constitutes a TN in the context of segmentation.

If your validation dataset contains a significant number of empty images (no tumors), consider separating the analysis for images with ROIs and without ROIs. For images without ROIs, if your model does not highlight any area, you could manually count these as TNs. However, this manual TN count won't be represented in the automated confusion matrix generated during validation, as the segmentation models essentially predict "presence" or "absence" explicitly through the segmentation mask.

I hope this clarifies your concerns. Engaging in medical image segmentation for your thesis is indeed a challenging yet rewarding endeavor. Best of luck, and don't hesitate to reach out should you have more inquiries!

**Reply by MarioPasc** [reply]
Hello!

Thank you very much for taking the time to respond. I have carefully considered your answer and compared it with my code.

I have a further question regarding the generation of a dataset from scratch. Since I am working with MRI images, which can have brightness values that exceed the np.uint8 range, I am developing my own data augmentation class. This has given me deep insight into how to generate the YOLOv8 dataset. However, I am having some trouble understanding how the model computes the training cache and validation cache files. Specifically, I am unclear on how YOLOv8 analyzes the provided dataset to label each sample as "background" (a .png in the train/val folder without a corresponding .txt label) or "image" (a .png in the train/val folder with a corresponding .txt file).

When I upload my dataset to the YOLOv8 model, here are the statistics it provides for my train folder:
```
train: Scanning (...) 2144 images, 2994 backgrounds, 0 corrupt: 100%|██████████| 5138/5138 [00:03<00:00, 1638.83it/s]
```
However, when I count the number of .txt files in my labels/train folder (FCD=1) and the number of .png images in images/train folder (FCD=1+FCD=0) and subtract these two numbers to get the theoretical number of "background" files, I find a discrepancy of nearly 800 images:
![dataset_distribution-BRGM](https://github.com/ultralytics/ultralytics/assets/120520768/3717a323-93f5-4bf7-bbb3-960310e9a40e)

I recognize that I have taken a naïve approach to analyzing my YOLOv8 generated dataset. Could you help me understand how the model analyzes the given dataset to generate the training and validation cache? This information might help me determine whether my issue lies in how I am computing my graphics for my generated dataset statistics, or if I am missing some files or mismatches in file names when generating my augmented data.

Thank you very much!. I appreciate your time and look forward to any guidance you can provide.

**Reply by pderrenger** [reply]
Hello!

Thanks for following up with your detailed observations! I'm here to help you navigate through your dataset nuances 🌟.

In YOLOv8, when we talk about "background" and "image" classifications during dataset preparation, it essentially comes down to the presence or absence of `.txt` label files for each `.png` image. An image is considered as having a "background" (no object of interest) if it lacks an accompanying `.txt` label file. Conversely, if there is a `.txt` file, even if it's empty, the image is treated differently by the caching mechanism.

Regarding the cache generation process, YOLOv8 inspects your dataset's structure, counting `.png` files and their corresponding `.txt` files (if present) to prepare training and validation caches. The discrepancy you're observing could potentially arise if there are:
- Misnamed or misplaced `.txt` files not aligning with their `.png` counterparts.
- Hidden files or folders accidentally counted.
- Divergences in how your custom data augmentation might manage or output `.txt` labels, especially if augmentation affects file naming or placement. 

To debug this:
- Verify the alignment of image file names with their respective label file names.
- Ensure there are no extraneous or hidden files in your folders that could be misinterpreted as additional data.
- Check for consistency in your data augmentation process, particularly how it outputs or modifies file names.

Example for aligning image and label paths:
```python
import os

image_folder = 'path/to/images'
label_folder = 'path/to/labels'

# Find missing or misaligned label files
for img_file in os.listdir(image_folder):
    label_file = img_file.replace('.png', '.txt')
    if not os.path.exists(os.path.join(label_folder, label_file)):
        print(f"Missing or misaligned label for {img_file}")
```

Let's ensure all images and their labels are correctly paired and accounted for. This might help resolve the mismatch you've found. I'm here if you have more questions or need further clarification. Keep exploring the depths of your data, and happy coding!

**Reply by MarioPasc** [reply]
Hi! 
Thank you for you guidance! I debugged my code and detected the error, now I get the same result for in my generated plot and in the yolov8 dataset analysis 

```
train: Scanning  (...)/train... 2932 images, 3200 backgrounds, 0 corrupt: 100%|██████████| 6132/6132 [00:04<00:00, 1444.63it/s
train: New cache created: /home/mariopasc/Python/Datasets/T2FLAIR_yolov8_dataset/labels/train.cache
val: Scanning (...)/val... 1055 images, 930 backgrounds, 0 corrupt: 100%|██████████| 1985/1985 [00:00<00:00, 6542.75it/s]
val: New cache created: /home/mariopasc/Python/Datasets/T2FLAIR_yolov8_dataset/labels/val.cache
```
 
![dataset_distribution_BRGM](https://github.com/ultralytics/ultralytics/assets/120520768/3b6b8f5c-4d3d-438d-b789-40c6d858741e)

I truly appreciate your dedication to supporting users like myself and your commitment to the continuous improvement of the Ultralytics projects :)

**Reply by pderrenger** [reply]
Hi there! 🌟

Fantastic news on squashing that bug and aligning your dataset statistics with the YOLOv8 analysis output! It’s always rewarding to see hard work pay off, and I’m thrilled we could be a part of your journey. Your kind words mean a lot to us, and we're committed to supporting our users and improving Ultralytics projects every day. 🚀

Happy to have helped, and don't hesitate to reach out if you dive into new challenges or have more questions in the future. Keep up the great work!

Happy coding! 🎉
